src/lib/modules/blockchain_process/proxy.js

```diff
@@ -1,156 +1,180 @@
   try {
-    jsonO = JSON.parse(reqBody);
+    if (string && typeof string === 'string') object = JSON.parse(string);
   } catch (e) {
-    return; // Request is not a json. Do nothing
-  }
-  if (jsonO.method === "eth_sendTransaction") {
-    commList[jsonO.id] = {
-      type: 'contract-log',
-      address: jsonO.params[0].to,
-      data: jsonO.params[0].data
-    };
-  } else if (jsonO.method === "eth_getTransactionReceipt") {
-    if (transactions[jsonO.params[0]]) {
-      transactions[jsonO.params[0]].receiptId = jsonO.id;
-      receipts[jsonO.id] = transactions[jsonO.params[0]].commListId;
-    }
+    console.error('Error parsing string as JSON', string);
+  } finally {
+    // eslint-disable-next-line no-unsafe-finally
+    return object;
   }
 };
 
-const parseResponse = function (ipc, resBody) {
-  let jsonO;
-  try {
-    jsonO = JSON.parse(resBody);
-  } catch (e) {
-    return; // Response is not a json. Do nothing
-  }
-  if (commList[jsonO.id]) {
-    commList[jsonO.id].transactionHash = jsonO.result;
-    transactions[jsonO.result] = {commListId: jsonO.id};
-  } else if (receipts[jsonO.id] && jsonO.result && jsonO.result.blockNumber) {
-    // TODO find out why commList[receipts[jsonO.id]] is sometimes not defined
-    if (!commList[receipts[jsonO.id]]) {
-      commList[receipts[jsonO.id]] = {};
+exports.serve = async (ipc, host, port, ws, origin) => {
+  const commList = {};
+  const receipts = {};
+  const transactions = {};
+
+  const trackRequest = (req) => {
+    if (!req) return;
+    try {
+      if (req.method === 'eth_sendTransaction') {
+        commList[req.id] = {
+          type: 'contract-log',
+          address: req.params[0].to,
+          data: req.params[0].data
+        };
+      } else if (req.method === 'eth_getTransactionReceipt') {
+        if (transactions[req.params[0]]) {
+          transactions[req.params[0]].receiptId = req.id;
+          receipts[req.id] = transactions[req.params[0]].commListId;
+        }
+      }
+    } catch (e) {
+      console.error('Error tracking request message', JSON.stringify(req));
     }
-    commList[receipts[jsonO.id]].blockNumber = jsonO.result.blockNumber;
-    commList[receipts[jsonO.id]].gasUsed = jsonO.result.gasUsed;
-    commList[receipts[jsonO.id]].status = jsonO.result.status;
-
-    if (ipc.connected && !ipc.connecting) {
-      ipc.request('log', commList[receipts[jsonO.id]]);
-    } else {
-      const message = commList[receipts[jsonO.id]];
-      ipc.connecting = true;
-      ipc.connect(() => {
-        ipc.connecting = false;
-        ipc.request('log', message);
-      });
+  };
+
+  const trackResponse = (res) => {
+    if (!res) return;
+    try {
+      if (commList[res.id]) {
+        commList[res.id].transactionHash = res.result;
+        transactions[res.result] = {commListId: res.id};
+      } else if (receipts[res.id] && res.result && res.result.blockNumber) {
+        // TODO find out why commList[receipts[res.id]] is sometimes not defined
```

Comment:
@michaelsbradleyjr is this something you still wanted to do, before this gets merged?

Reply:
The functions `trackRequest` and `trackResponse` were inherited from the previous implementation, with only minor changes: https://github.com/embark-framework/embark/blob/master/src/lib/modules/blockchain_process/proxy.js#L12-L67.

My understanding of what they're looking for and why isn't too deep; in this PR I mainly focused on making sure that JSON-RPC messages are reliably being parsed from the inbound and outbound streams. I think it's best, at this point, to not introduce further modifications and then in another PR/s we could introduce improvements. Thoughts?

Reply:
Ah understood. Sorry I missed that. Fair enough :)